src/stream/src/common/table/state_table.rs

```diff
@@ -805,13 +943,38 @@ where
         } else {
             Some(self.pk_serde.prefix(1))
         };
-        let watermark_suffix = watermark.map(|watermark| {
+
+        let should_clean_watermark = match watermark {
+            Some(ref watermark) => {
+                if USE_WATERMARK_CACHE && self.watermark_cache.is_synced() {
+                    if let Some(key) = self.watermark_cache.lowest_key() {
+                        watermark.as_scalar_ref_impl().default_cmp(&key).is_ge()
+                    } else {
+                        // Watermark cache is synced,
+                        // And there's no key in watermark cache.
+                        // That implies table is empty.
+                        // We should not clean watermark.
+                        false
+                    }
+                } else {
+                    // Either we are not using watermark cache,
+                    // Or watermark_cache is not synced.
+                    // In either case we should clean watermark.
+                    true
+                }
+            }
+            None => false,
+        };
+
+        let watermark_suffix = watermark.as_ref().map(|watermark| {
             serialize_pk(
-                row::once(Some(watermark)),
+                row::once(Some(watermark.clone())),
                 prefix_serializer.as_ref().unwrap(),
             )
         });
-        if let Some(watermark_suffix) = watermark_suffix && let Some(first_byte) = watermark_suffix.first() {
+
+        // Compute Delete Ranges
+        if should_clean_watermark && let Some(watermark_suffix) = watermark_suffix && let Some(first_byte) = watermark_suffix.first() {
```

Comment:
I remember someone said we may do point delete if the `state_cleaning_watermark` falls into the cache, did that reach a conclusion?

Reply:
Oh I found it, need benchmark, yes.